Source/utils/WeylExtraction.impl.hpp

```diff
@@ -81,62 +92,82 @@ inline void WeylExtraction::execute_query(
 }
 
 //! integrate over a spherical shell with given harmonics
-inline std::array<double, 2>
+inline std::vector<double>
 WeylExtraction::integrate_surface(int es, int el, int em,
                                   const double *m_state_ptr_re,
                                   const double *m_state_ptr_im) const
 {
+    CH_TIME("WeylExtraction::integrate_surface");
     int rank;
 #ifdef CH_MPI
     MPI_Comm_rank(Chombo_MPI::comm, &rank);
 #else
     rank = 0;
 #endif
-    std::array<double, 2> integral = {0.0, 0.0};
+    std::vector<double> integral(2 * m_params.num_extraction_radii, 0.);
 
     // only rank 0 does the integral, but use OMP threads if available
     if (rank == 0)
     {
         // integrate the values over the sphere (normalised by r^2)
+        // for each radius
         // assumes spacings constant, uses trapezium rule for phi and rectangles
         // for theta  note we don't have to fudge the end points for phi because
         // the function is periodic  and so the last point (implied but not part
         // of vector) is equal to the first point
-#pragma omp parallel for
-        for (int iphi = 0; iphi < m_params.num_points_phi; ++iphi)
+#ifdef _OPENMP
+    #pragma omp parallel for collapse(2) default(none) \
+            shared(es, el, em, m_state_ptr_re, m_state_ptr_im, integral)
+#endif
+        for (int iradius = 0; iradius < m_params.num_extraction_radii;
+             ++iradius)
         {
-            double phi = iphi * 2 * M_PI / m_params.num_points_phi;
-            double inner_integral_re = 0;
-            double inner_integral_im = 0;
-            for (int itheta = 0; itheta < m_params.num_points_theta; itheta++)
+            for (int iphi = 0; iphi < m_params.num_points_phi; ++iphi)
```

Comment:
I guess with the existing OpenMP loop, it's hard to avoid the performance-hitting `atomic` directives. However, from my look at the Chombo timer outputs, I'm not sure much time is spent in this loop anyway. I did consider using OpenMP 4.5 C++ array reductions but I didn't think it was worth changing all the `std::vector`s to conventional arrays or using user-defined reductions given the minimal potential gain (and also, I think we decided to avoid using anything newer than OpenMP 4.0 for better compiler compatibility).

Reply:
Yes, totally, we should either use the atomic or leave out the radius loop and use standard reductions. I tested both versions compiled with g++-7 just now and it seems they are equally fast so it is up to you :-P.